Schematron support #75

vincentml opened the pull request:
For feature request #75.

This pull request adds support for testing Schematron with XSpec. See the file tutorial/schematron/schematron.txt for documentation which could be added to the wiki.

cirulls commented:
@vincentml : thanks for this pull request, it's awesome to see the XSpec community extending XSpec! Thanks also for providing documentation, a tutorial, a test, and a schema. This is going to be very useful when the new feature is released. 

I noticed that some tests are failing in [Travis](https://travis-ci.org/xspec/xspec/builds/210322529) and [AppVeyor](https://ci.appveyor.com/project/xspec/xspec/build/1.0.150). The failing test on Travis is  trivial: since you introduced a new command line option, the test for printing the usage needs to be amended. AppVeyor has got eight failing tests which I have not assessed. I may write some documentation on how to run the test suite locally so that contributors can check it on their machine before starting a pull request. 

@tofi86: you wrote you want to have a look in https://github.com/xspec/xspec/issues/75#issuecomment-285978922. Would you be able to review @vincentml 's pull request? This involves trying it out, checking that it is working as described,  and making sure you can follow the documentation and the tutorial @vincentml wrote. 

@vincentml, just an idea for a (future) enhancement: I noticed you committed the Schematron code in your pull request. I wonder if it would be possible to import it from [GitHub](https://github.com/Schematron/schematron) and set up a variable to define the location of Schematron. This is similar to what we do with Saxon and allows to pick up the latest version.

AirQuick commented:
> AppVeyor has got eight failing tests

They come from a missing `if defined`. I'll put some review comments for that and other (trivial) Windows things.

AirQuick commented on bin/xspec.bat:
```diff
@@ -120,6 +121,9 @@ rem ##
         set XQUERY=1
+    ) else if "%WIN_ARGV%"=="-s" (
+        set SCHEMATRON=1
+        set XSLT=1
```
* `SCHEMATRON` should be reset at `:win_reset_options`.
  * The purpose of `:win_reset_options` is to negate the environmental variables when they are already set outside the batch.

* `XSLT` should not be set here.
  * Setting `XSLT` here makes it impossible to tell "-s and -t are mutually exclusive".
  * `XSLT` is set [later](https://github.com/xspec/xspec/blob/f4811a5976c637210e651f5b46d962f38d183f9b/bin/xspec.bat#L310) anyway.

AirQuick commented on bin/xspec.bat:
```diff
@@ -260,6 +264,15 @@ call :win_reset_options
+    call :usage "-s and -q are mutually exclusive"
+    exit /b 1
+)
+
```
Checking the case of "-s and -t are mutually exclusive" is missing

vincentml replied:
Fixed, check was added for -s and -t are mutually exclusive

AirQuick replied:
Thanks. I was too lazy to verify it myself, so I pushed Travis/AppVeyor tests for it like [this](https://github.com/AirQuick/xspec/commit/77874e03908a8b10610e0057dc217be853433972).
Can you consider adding something like that to this pull request?

AirQuick commented on bin/xspec.bat:
```diff
@@ -372,6 +385,48 @@ rem ## compile the suite #######################################################
 rem
 
+rem # get URI to Schematron file and phase from the SchUT file
+if defined SCHEMATRON (
```
I see a series of `if defined SCHEMATRON` is inevitable, but it looks error-prone. It would be nice to do `goto` and skip the bulk of the lines. That will
* simplify the code
* prevent mistakes like the current test [failure](https://ci.appveyor.com/project/xspec/xspec/build/1.0.150/job/sico8n9i8ogkdk9r#L108)

Though `goto` is in general not a good habit, I think for this particular case the end justifies the means.

AirQuick commented on bin/xspec.bat:
```diff
@@ -505,6 +560,8 @@ if defined COVERAGE (
     rem %OPEN% "%HTML%"
 )
 
+if defined SCHEMATRON del %SCHUT%
```
As a general precaution, surrounding quotes for `%SCHUT%` would be necessary.

tofi86 commented:
> @tofi86: you wrote you want to have a look in #75 (comment). Would you be able to review @vincentml 's pull request? This involves trying it out, checking that it is working as described, and making sure you can follow the documentation and the tutorial @vincentml wrote.

I cannot promise anything at the moment as I'm having my wedding party on Saturday 🎈 but I'll try to have a look the week after on my flight to Toronto to #ebookcraft or somewhen after that...
However, if someone else is having more time to do proper review, please go ahead!

vincentml commented:
@AirQuick Thanks for the feedback! I've made the changes that you suggested in xspec.bat and updated the pull request. It looks like only one test is failing now and that is on printing the usage information.

vincentml commented:
I was able to fix the tests for the usage message. There is still one test failing in AppVeyor which I have not been able to reproduce locally. My guess is that it has something to do with the relative file paths in schut-to-xspec.xspec

AirQuick commented:
@vincentml , Thanks for the update. I like [this idea](https://github.com/vincentml/xspec/blob/884f576ef8aa04ca54b2dc1f7ce39fb9e66cf290/bin/xspec.bat#L439) in particular.
For the `schut-to-xspec.xspec` failure, somehow I got a repro locally. I'll look into it.
To help investigation like this, AppVeyor should have provided more information on failure. I'll file an issue for it.

AirQuick commented on test/run-xspec-tests.sh:
```diff
@@ -23,8 +23,13 @@
-do ../bin/xspec.sh $xspectest &> result.log; 
+for xspectest in *.xspec *.schut; 
+do 
+    if test "${xspectest#.schut}" = "$xspectest"; then
```
[Injecting `echo`](https://github.com/AirQuick/xspec/commit/4dcefff664b60b0c761375b118f2d99f5dc7ee22) for testing purposes suggests that [every file](https://travis-ci.org/AirQuick/xspec/jobs/211303084#L471) enters the Schematron route. Looks suspicious.

AirQuick commented on test/run-xspec-tests.sh:
```diff
@@ -23,8 +23,13 @@
+    else 
+        ../bin/xspec.sh $xspectest &> result.log;
+    fi
     if grep -q ".*failed:\s[1-9]" result.log || grep -q -E "\*+\sError\s(running|compiling)\sthe\stest\ssuite" result.log;
```
This does not catch failures in the Schematron phase, such as "*** Error compiling the schematron on step 1".
(That is what's happening on Travis right now, I guess.)

vincentml replied:
Fixed: Error messages related to Schematron should now be caught

AirQuick commented:
> There is still one test failing in AppVeyor which I have not been able to reproduce locally.

The failure happens because unfortunately a stubborn someone has broken the master branch for his own sake. Since the failure does not happen on [a healthy branch](https://ci.appveyor.com/project/AirQuick/xspec/build/119-pr-105/job/0k3dgc1q2nx3s830#L126), I think we can ignore it for now.

cirulls commented:
After merging both #101 and #103 tests on branch ```master``` are back to green on on [Travis](https://travis-ci.org/xspec/xspec/builds/211363074) and [AppVeyor](https://ci.appveyor.com/project/xspec/xspec/build/1.0.163). 

Could you please rebase your branch and check that test is now passing?

AirQuick commented on test/run-xspec-tests.sh:
```diff
@@ -4,7 +4,7 @@
 #         
 #   DESCRIPTION:  This script is a runner for XSpec test suite inside test directory
-#		  it runs all the tests with extension .xspec in the test directory
+#		  it runs all the tests with extension .xspec or .schut in the test directory
```
Nitpick. Is this edit still true?

vincentml replied:
Thanks for spotting that. Fixed.

AirQuick commented on bin/xspec.sh:
```diff
@@ -158,14 +159,34 @@ while echo "$1" | grep -- ^- >/dev/null 2>&1; do
+                exit 1
+            fi
+            SCHEMATRON=1;
+            XSLT=1;;
```
I'm not sure but maybe this `XSLT` confuses the script into thinking "-s and -q" is "-t -and -q". See [log](https://travis-ci.org/AirQuick/xspec/jobs/211662119#L446-L449).

vincentml replied:
I had fixed this in xspec.bat but missed it in xspec.sh. Fixed now, and I added tests for the error messages about mutually exclusive parameters -s -t -q.

AirQuick commented on test/run-xspec-tests.cmd:
```diff
@@ -64,3 +69,12 @@ rem
+
+:is_schematron
+    set var=%1
+    if not "%var%"=="%var:schematron=%" (
```
Would be nice to align it with the shell script: Something like `    if "%var:~0,10%"=="schematron" (`

vincentml replied:
This, with the change in the in the previous comment, is better. Thanks!

AirQuick commented on test/run-xspec-tests.cmd:
```diff
@@ -39,12 +39,17 @@ for %%I in (*.xspec) do (
     rem Run
     rem
-    "%COMSPEC%" /c ..\bin\xspec.bat "%%~I" > "%RESULT_FILE%" 2>&1
+    call :is_schematron %%~nI%
```
Just in case, wrapping by `call :is_schematron "%%~nI"` and unwrapping by `set var=%~1` would be safer.

vincentml replied:
Thanks! I didn't know about using `%~1` to unwrap quotes

AirQuick commented on test/run-xspec-tests.cmd:
```diff
@@ -39,12 +39,17 @@ for %%I in (*.xspec) do (
     rem Run
     rem
-    "%COMSPEC%" /c ..\bin\xspec.bat "%%~I" > "%RESULT_FILE%" 2>&1
+    call :is_schematron "%%~nI%"
```
This doesn't have adverse affect right now, but the last `%` should be removed: `"%%~nI"`

Thanks for all the housekeeping you've taken!

vincentml replied:
Now fixed

vincentml commented:
I pushed a commit but it is not shown in the pull request; maybe GitHub is delayed.

AirQuick commented:
@vincentml ,
Thank you for the further improvements and also for adding the tests to make sure everything's ok!
I mirrored your branch and one of the tests failed due to a subtle condition (whether the output directory exists or not). See [the log](https://ci.appveyor.com/project/AirQuick/xspec/build/139-pr-105/job/mf6pd84h2ohsfa4l#L138).
Similar issues did happen for the other tests. That's why the output directories are handled at `setup` and `teardown`.

AirQuick commented:
Just noticed a suspicious [message](https://ci.appveyor.com/project/AirQuick/xspec/build/139-pr-105/job/mf6pd84h2ohsfa4l#L159): `[20]The system cannot find the path specified.`
Maybe the forward slashes in [`xspec-bat.cmd line#300`](https://github.com/vincentml/xspec/blob/b0dc03d1a9a8b33a6afb044ec7a056c6e8aee827/test/xspec-bat.cmd#L300) should have been backslashes?

vincentml commented:
@AirQuick thanks for pointing out the problems with the file path and output directory. These are now fixed. A merge conflict occurred blocking my last push, so I've rebased my pull request branch and resolved the conflict.

AirQuick commented on src/compiler/generate-xspec-tests.xsl:
```diff
@@ -87,6 +87,7 @@
     </template>
     <!-- Compile the top-level scenarios. -->
     <xsl:call-template name="x:compile-scenarios"/>
+    <xsl:copy-of select="document('')/xsl:stylesheet/xsl:function[contains(@name, 'schematron-location-')]"/>
```
Just curious. Why embedding the functions directly rather than creating `xsl:import`?

vincentml replied:
I should have done the easier thing first. Now fixed. Thanks for catching that.

AirQuick commented on test/xspec-bat.cmd:
```diff
@@ -301,6 +331,30 @@ setlocal
 endlocal
 
+setlocal
+    call :setup "executing the Saxon XProc harness generates a report with UTF-8 encoding"
```
[Dup](https://github.com/vincentml/xspec/blob/de4a2ad60bef1e2b6542882fca95869ca3b1549b/test/xspec-bat.cmd#L290)? Maybe a remnant of conflicts.

vincentml replied:
Now fixed. Thanks for catching that.
